Preserve freq in TimedeltaIndex subtraction

Chenghaoxiang opened the pull request:
- [x] closes #62094  
- [x]  Tests added and passed if fixing a bug or adding a new feature
- [x] All code checks passed.
- [x] Added type annotations to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Chenghaoxiang committed:
Preserve freq in TimedeltaIndex subtraction

jbrockmendel commented on pandas/core/arrays/datetimelike.py:
```diff
@@ -1089,6 +1089,12 @@ def _get_arithmetic_result_freq(self, other) -> BaseOffset | None:
+        elif (
+            lib.is_np_dtype(self.dtype, "M")
+            and isinstance(other, Timestamp)
+            and cast("DatetimeArray", self).tz == other.tz
```
Tz check should be unnecessary; this should only work for Day freqs.

Chenghaoxiang replied:
ok, on it.

jbrockmendel commented:
Can you add a test for the fix

Chenghaoxiang commented:
Certainly. I've added a test case and updated the PR. Please let me know if anything else is needed.

jbrockmendel commented on pandas/tests/arrays/test_datetimelike.py:
```diff
@@ -1356,3 +1356,40 @@ def test_from_pandas_array(dtype):
+    ],
+)
+@pytest.mark.parametrize("periods", [1, -2, 0])
+def test_shift_after_dti_sub_timestamp_various_inputs(dti, ts, periods):
```
You don’t need all this. Just one test in tests.indexes.dattimes.arithmetic for datetimeindex-timestamp retaining freq

Chenghaoxiang commented:
Thanks for the guidance! I've updated the PR to restrict the logic to `Day` frequency and replaced the test with a single, focused regression test in the location you recommended.
Appreciate the patient review!

Chenghaoxiang committed:
Merge branch 'main' into new-feature

Chenghaoxiang commented:
Hi  @jbrockmendel, I just wanted to gently check in on this. Was there an error on my part that I missed? I'd appreciate it if you could get back to me. Thank you.

Chenghaoxiang committed:
Merge branch 'main' into new-feature

Chenghaoxiang commented:
My pleasure. Happy to contribute!
